File: cpp/logger/Logger.cpp
/**
 * Copyright 2004-present, Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "Logger.h"

#include <algorithm>
#include <chrono>
#include <cstring>
#include <stdexcept>

#include <profilo/util/common.h>

namespace facebook {
namespace profilo {

using namespace entries;

Logger::EntryIDCounter& Logger::getGlobalEntryID() {
  static EntryIDCounter global_instance{kDefaultInitialID};
  return global_instance;
}

Logger::Logger(logger::TraceBufferProvider provider, EntryIDCounter& counter)
    : entryID_(counter), logger_(provider) {}

int32_t Logger::writeBytes(
    EntryType type,
    int32_t arg1,
    const uint8_t* arg2,
    size_t len) {
  if (len > kMaxVariableLengthEntry) {
    throw std::overflow_error("len is bigger than kMaxVariableLengthEntry");
  }
  if (arg2 == nullptr) {
    throw std::invalid_argument("arg2 is null");
  }

  return write(BytesEntry{
      .id = 0,
      .type = type,
      .matchid = arg1,
      .bytes = {
          .values = const_cast<uint8_t*>(arg2),
          .size = static_cast<uint16_t>(len)}});
}

} // namespace profilo
} // namespace facebook

File: java/main/com/facebook/profilo/logger/Logger.java
/**
 * Copyright 2004-present, Facebook, Inc.
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.profilo.logger;

import com.facebook.profilo.core.ProfiloConstants;
import com.facebook.profilo.core.TraceEvents;

public final class Logger {

  private static volatile boolean sInitialized;

  public static void initialize() {
    sInitialized = true;
  }

  public static final int SKIP_PROVIDER_CHECK = 1 << 0;
  public static final int FILL_TIMESTAMP = 1 << 1;
  public static final int FILL_TID = 1 << 2;

  public static int writeStandardEntry(
      int provider,
      int flags,
      int type,
      long timestamp,
      int tid,
      int arg1 /* callid */,
      int arg2 /* matchid */,
      long arg3 /* extra */) {
    if (sInitialized && ((flags & SKIP_PROVIDER_CHECK) != 0 || TraceEvents.isEnabled(provider))) {
      return BufferLogger.writeStandardEntry(null, flags, type, timestamp, tid, arg1, arg2, arg3);
    } else {
      return ProfiloConstants.TRACING_DISABLED;
    }
  }

  public static int writeBytesEntry(
      int provider, int flags, int type, int arg1 /* matchid */, String arg2 /* bytes */) {
    if (sInitialized && ((flags & SKIP_PROVIDER_CHECK) != 0 || TraceEvents.isEnabled(provider))) {
      return BufferLogger.writeBytesEntry(null, flags, type, arg1, arg2);
    } else {
      return ProfiloConstants.TRACING_DISABLED;
    }
  }
}
